### PR TITLE
feat(pam): add local dev stack for seeding PAM resources

### PR DESCRIPTION
**Pull request by saifsmailbox98**
# Description 📣

Spins up local databases and SSH servers and registers them in Infisical as PAM resources. Useful for testing PAM features and reproducing customer issues.

## Type ✨

- [ ] Bug fix
- [x] New feature
- [ ] Improvement
- [ ] Breaking change
- [ ] Documentation

# Tests 🛠️

```sh
# Here's some code block to paste some code snippets
```

---

- [x] I have read the [contributing guide](https://infisical.com/docs/contributing/getting-started/overview), agreed and acknowledged the [code of conduct](https://infisical.com/docs/contributing/getting-started/code-of-conduct). 📝

### CODE REVIEW

File: dev/pam/resources/postgres/init/seed.sql [hunk 1]
```diff
@@ -0,0 +1,1629 @@
...
+
+DO $$
+DECLARE
+    pids UUID[] := ARRAY(SELECT id FROM products);
+    p UUID;
+BEGIN
+    FOR i IN 1..500 LOOP
+        p := pids[1 + floor(random() * array_length(pids, 1))::int];
+        INSERT INTO price_history (product_id, price, changed_at, reason)
+        VALUES (
```
Comment (bernie-g) [inline dev/pam/resources/postgres/init/seed.sql:433]: ignore placement of comment

What are your thoughts on adding some really big tables to these seeds? Would be useful to test edge cases for really long queries/timeouts

File: dev/pam/resources/redis/users.acl [hunk 2]
```diff
@@ -0,0 +1,2 @@
+user default off
```
Comment (bernie-g) [inline dev/pam/resources/redis/users.acl:1]: ignore comment placement

Can we add some keys to redis? Seems a little weird that it's the only one that doesn't have any seeded data

**Review by bernie-g — COMMENTED** [review]
ignore placement of comment

File: dev/pam/resources/redis/users.acl (hunk 2 above)
Comment (saifsmailbox98) [reply dev/pam/resources/redis/users.acl:1]: added some data to redis as well

File: dev/pam/resources/postgres/init/seed.sql (hunk 1 above)
Comment (saifsmailbox98) [reply dev/pam/resources/postgres/init/seed.sql:433]: i actually thought about this, but it would take a lot of space when they run it, and not everyone is working on postgres with huge amounts of data, so it's probably not a good time to add it. we can probably do that once we feel the the need for it, or maybe just add a seed script that can be run manually.

for now, the aim is to keep it simple and lightweight so it's easy for others to pick up, and to see whether it's even worth having in the first place.

File: dev/pam/resources/mssql/entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+IF NOT EXISTS (SELECT name FROM sys.database_principals WHERE name = N'$USR')
+    CREATE USER [$USR] FOR LOGIN [$USR];
+
+ALTER ROLE db_owner ADD MEMBER [$USR];
+"
+
+# Run init scripts
+for f in /docker-entrypoint-initdb.d/*.sql; do
+    if [ -f "$f" ]; then
```
[the remark this answers is not in the document]
Comment (saifsmailbox98) [reply dev/pam/resources/mssql/entrypoint.sh]: fixed

File: dev/pam/Makefile [hunk 4]
```diff
@@ -0,0 +1,43 @@
+SHELL := /bin/bash
+
+.PHONY: up down clean setup info
+
+up:
```
[the remark this answers is not in the document]
Comment (saifsmailbox98) [reply dev/pam/Makefile:5]: added the friendly message here as well

File: dev/pam/setup.sh [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+  if (( ${#rows[@]} > 0 )); then
+    for r in "${rows[@]}"; do echo "$r"; done
+  else
+    echo "(no services currently running — 'make up' brings them up)"
+  fi
+  echo ""
+}
+
+if (( PRINT_INFO_ONLY )); then
+  print_info
```
[the remark this answers is not in the document]
Comment (saifsmailbox98) [reply dev/pam/setup.sh:99]: using local01 everywhere

File: dev/pam/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      SSH_USER: infisical
+      SSH_MODE: password
+      SSH_PASSWORD: Infisical@123
+    ports:
+      - "127.0.0.1:55422:22"
+
+  ssh-server-key:
+    build:
+      context: ./resources/ssh-server
+    profiles: [ssh-key]
```
[the remark this answers is not in the document]
Comment (saifsmailbox98) [reply dev/pam/docker-compose.yml:133]: we can add this later, no need to complicate it for now

File: dev/pam/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+      MONGO_INITDB_ROOT_USERNAME: infisical
+      MONGO_INITDB_ROOT_PASSWORD: Infisical@123
+      MONGO_INITDB_DATABASE: infisical
+    ports:
+      - "127.0.0.1:55437:27017"
+    volumes:
+      - ./resources/mongodb/init:/docker-entrypoint-initdb.d
+      - pam_mongodata:/data/db
+    healthcheck:
+      test: ["CMD", "mongosh", "--eval", "db.adminCommand('ping')"]
```
[the remark this answers is not in the document]
Comment (saifsmailbox98) [reply dev/pam/docker-compose.yml:77]: we can add this later, no need to complicate it for now

File: dev/pam/resources/postgres/init/seed.sql (hunk 1 above)
Comment (bernie-g) [reply dev/pam/resources/postgres/init/seed.sql:433]: Hm ok fair enough

File: dev/pam/setup.sh [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+AUTH_HEADER="Authorization: Bearer ${INFISICAL_TOKEN}"
+PREFIX="${RESOURCE_PREFIX}-$(date +%F)"
+
+# -----------------------------------------------------------------------------
+# Gateway ID — required by the Infisical PAM create-resource endpoints
+# -----------------------------------------------------------------------------
+if [[ -z "${INFISICAL_GATEWAY_ID:-}" ]]; then
+  echo "INFISICAL_GATEWAY_ID is not set. Fetching gateways from ${INFISICAL_DOMAIN} ..."
+  gw_resp=$(curl -sS -H "$AUTH_HEADER" "${INFISICAL_DOMAIN}/api/v1/gateways/")
```
Comment (bernie-g) [inline dev/pam/setup.sh]: Why does this hit v1?

Comment (saifsmailbox98) [reply dev/pam/setup.sh]: this should have been v2, but removed the gateway id fetching completely

making it more user friendly could add more points of failure when using and harder to debug

File: dev/pam/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,149 @@
+services:
+  postgres:
+    image: postgres:17-alpine
+    profiles: [postgres]
+    environment:
+      POSTGRES_DB: infisical
+      POSTGRES_USER: infisical
+      POSTGRES_PASSWORD: Infisical@123
```
Comment (bernie-g) [inline dev/pam/docker-compose.yml:8]: Thoughts on standardizing all the passwords to just `infisical`? That's the password for the local infisical db and it would be nice to keep it consistent + extra simple to remember

Comment (saifsmailbox98) [reply dev/pam/docker-compose.yml:8]: I started with `infisical` initially, but mssql had a different requirement: https://learn.microsoft.com/en-us/sql/relational-databases/security/password-policy?view=sql-server-ver17#password-complexity

the policy can be turned off though, but doing that might lead to unintended behavior for someone testing something related to auth, etc.

could have used `infisical` for others though, but that might create more confusion, so changed all of them to `Infisical@123`, another advantage of this is that if you're looking for places where password/credentials are defined you can do so with a search for `Infisical@123` - `infisical` might match a lot more places.

let me know what you think

Comment (bernie-g) [reply dev/pam/docker-compose.yml:8]: Ah ok in that case I think Infisical@123 is fine for consistency. Probably not worth the hassle

**Review by bernie-g — APPROVED** [review]
